### PR TITLE
feat: add support for trust grants that can issue tokens for any subject

### PR DESCRIPTION
**Pull request by jagobagascon**
Previously, a trust relationship had to be setup for every subject
before the issuer could sign a JWT token for it. This change will allow
setting up token services that can issue tokens with any value in the
subject field.

## Related issue(s)
#2930

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](docs/docs).

### REST api
This PR adds a new field to the trust creation request called `allow_any_subject`, and as long as this flag is set to false (or unset), the trust relationship creation service will behave just like it did before this change: the `subject` field will be required.

```json
{
  "issuer": "https://jwt-idp.example.com",
  "subject": "mike@example.com",
  "...": "..."
}
```

When `allow_any_subject` is set to true, then the issuer will be allowed to sign tokens for any subject. Also, as a safety measure, the `subject` field MUST be empty (or unset), otherwise the validation will fail. This is done to avoid creating an issuer that can sign tokens for any subject by mistake.

```json
{
  "issuer": "https://jwt-idp.example.com",
  "allow_any_subject": true,
  "...": "..."
}

```

### Database

When `allow_any_subject` is set to true, then the subject field is stored with an empty string value in the database. This allows us to keep (and benefit from) the `UNIQUE` constraint that the `hydra_oauth2_trusted_jwt_bearer_issuer` table has, requires no changes to the database schema and makes the database query used to find an issuer's public key really simple: `subject = ? or subject = ''`.

### CODE REVIEW

**Comment by jagobagascon** [comment]
I've seen that the documentation is now in another repository, and I'm not sure if I should update the docs now or wait for this PR to be merged in case there's any changes.

**Comment by jagobagascon** [comment]
Hi @aeneasr, I'm not sure why e2e tests are failing. I just ran them locally and everything seems to be working fine. The error it's throwing is 
```
"error": "missing_required_parameter",
"error_description": "One of the required parameters is missing. Check your request parameters. Field 'subject' is required."
```

But that string has been deleted from the code, it's almost as if it was running on an older binary. Could it be related to the `actions/cache@v2` GitHub action used in the end-to-end tests setup? I'm a bit confused right now.

File: persistence/sql/persister_grant_jwk.go [hunk 1]
```diff
@@ -90,7 +90,7 @@ func (p *Persister) GetPublicKey(ctx context.Context, issuer string, subject str
 	var data trust.SQLData
 	query := p.Connection(ctx).
 		Where("issuer = ?", issuer).
-		Where("subject = ?", subject).
+		Where("subject = ? or subject = ''", subject).
```
Comment (aeneasr) [inline persistence/sql/persister_grant_jwk.go:93]: I would prefer a separate SQL key `allow_any_subject` and then: `WHERE subject = ? OR allow_any_subject IS TRUE`

**Review by aeneasr — CHANGES_REQUESTED** [review]
Thank you! Sorry for the late review! This looks really good. The one thing I would like to change is that we store this "any" relationship explicitly in the database. An empty string could get into the DB with a number of things (e.g. broken migrations, rollbacks, ...) and we really need to prevent that this feature is enabled "on accident". Therefore, please add a SQL column to store this state :)

File: persistence/sql/persister_grant_jwk.go (hunk 1 above)
Comment (jagobagascon) [reply persistence/sql/persister_grant_jwk.go]: Okay, it's also simpler and easier to understand!

**Comment by glerchundi** [comment]
Thanks @jagobagascon for your time & effort 👏 👏

**Review by aeneasr — APPROVED** [review]
Thank you so much! This looks great!

File: persistence/sql/persister_grant_jwk.go [hunk 2]
```diff
@@ -90,7 +90,7 @@ func (p *Persister) GetPublicKey(ctx context.Context, issuer string, subject str
 	var data trust.SQLData
 	query := p.Connection(ctx).
 		Where("issuer = ?", issuer).
-		Where("subject = ?", subject).
+		Where("subject = ? OR allow_any_subject IS TRUE", subject).
```
Comment (aeneasr) [inline persistence/sql/persister_grant_jwk.go:93]: Actually, could you please add a test for the persister that checks whether this works? :)

File: persistence/sql/persister_grant_jwk.go [hunk 3]
```diff
@@ -108,7 +108,7 @@ func (p *Persister) GetPublicKeys(ctx context.Context, issuer string, subject st
 	grantsData := make([]trust.SQLData, 0)
 	query := p.Connection(ctx).
 		Where("issuer = ?", issuer).
-		Where("subject = ?", subject)
+		Where("subject = ? OR allow_any_subject IS TRUE", subject)
```
Comment (aeneasr) [inline persistence/sql/persister_grant_jwk.go:111]: Same as above

File: persistence/sql/persister_grant_jwk.go [hunk 4]
```diff
@@ -138,7 +138,7 @@ func (p *Persister) GetPublicKeyScopes(ctx context.Context, issuer string, subje
 	var data trust.SQLData
 	query := p.Connection(ctx).
 		Where("issuer = ?", issuer).
-		Where("subject = ?", subject).
+		Where("subject = ? OR allow_any_subject IS TRUE", subject).
```
Comment (aeneasr) [inline persistence/sql/persister_grant_jwk.go:141]: Same as above

**Review by aeneasr — CHANGES_REQUESTED** [review]

**Comment by aeneasr** [comment]
@grantzvolsky please take note that this PR introduces a schema change. We'll squash merge it so that it's just one commit which hopefully will make it easier to rebase.

File: persistence/sql/persister_grant_jwk.go (hunk 2 above)
Comment (jagobagascon) [reply persistence/sql/persister_grant_jwk.go:93]: Of course!

**Comment by jagobagascon** [comment]
I added the tests as requested.

**Review by aeneasr — COMMENTED** [review]
Thank you so much for your great work @jagobagascon ! I added a small test, and now this is ready to be merged! But, one last ask - could you please update the document https://github.com/ory/docs/edit/master/docs/hydra/guides/oauth2-grant-type-jwt-bearer.mdx to include your changes so that others know about this feature? Thank you so much!

**Comment by jagobagascon** [comment]
Thank you @aeneasr! I updated the docs as requested: https://github.com/ory/docs/pull/764

**Comment by aeneasr** [comment]
🥳

**Comment by vinckr** [comment]
Hello @jagobagascon 
Congrats on merging your first PR in Ory 🎉 !
Your contribution will soon be helping secure millions of identities around the globe 🌏. 
As a small token of appreciation we send all our first time contributors a gift package to welcome them to the community. 
[Please drop me an email and I will forward you the form to claim your Ory swag!](mailto:vincent@ory.sh)

**Comment by jagobagascon** [comment]
Thank you @vinckr but I already received my gift for helping on a previous PR 😊.
Also thanks to @aeneasr for all the help and effort put on this. I can't wait to add this feature to our project!

If it's not too much to ask, are there any plans for the next Hydra release anytime soon?

**Comment by glerchundi** [comment]
Great job to all!! Cannot wait to see this working in prod! 🎉 🎉

**Pull request merged**